### PR TITLE
🚀 Adding interface for Ctx

### PR DESCRIPTION
**Pull request by thomasvvugt**
**Please provide enough information so that others can review your pull request:**
Instead of using a `struct` type, I suggest we implement an `interface` type for the context. By using an interface, we provide more expandability to our end-users for them to modify or embed our Ctx with custom logic.

**Explain the *details* for making this change. What existing problem does the pull request solve?**
This PR is related to `🚀 Extending Ctx using an interface`, see #691

### CODE REVIEW

**Comment by thomasvvugt** [comment]
Once merged, we might want to move to 1.14 as this PR has some impacting changes, such as:
- Removing the chainability of `ctx.Status()` and `ctx.Type()`
- Changing the Handler type from `func(*Ctx)` to `func(Ctx)`, which would require a lot of re-writing for our users

File: ctx.go [hunk 1]
```diff
@@ -28,22 +28,263 @@ import (
...
+		Write(bodies ...interface{})
+
+		// XHR returns a Boolean property, that is true, if the request's X-Requested-With header field is XMLHttpRequest,
+		// indicating that the request was issued by a client library (such as jQuery).
+		XHR() bool
+	}
+
+	// Context represents the Context which hold the HTTP request and response.
+	// It has methods for the request query string, parameters, body, HTTP headers and so on.
+	Context struct {
```
Comment (kiyonlin) [inline ctx.go:273]: What about making Context struct private?

File: ctx.go [hunk 2]
```diff
@@ -960,62 +1211,59 @@ func (ctx *Ctx) Subdomains(offset ...int) []string {
...
-func (ctx *Ctx) Status(status int) *Ctx {
-	ctx.Fasthttp.Response.SetStatusCode(status)
-	return ctx
+func (ctx *Context) Status(status int) {
+	ctx.Fasthttp().Response.SetStatusCode(status)
 }
 
 // Type sets the Content-Type HTTP header to the MIME type specified by the file extension.
-func (ctx *Ctx) Type(extension string, charset ...string) *Ctx {
+func (ctx *Context) Type(extension string, charset ...string) {
```
Comment (ReneWerner87) [inline ctx.go:1224]: why did you remove this return value here ? is this not implementable in the interface ?

File: ctx.go [hunk 3]
```diff
@@ -960,62 +1211,59 @@ func (ctx *Ctx) Subdomains(offset ...int) []string {
...
+func (ctx *Context) Stale() bool {
 	return !ctx.Fresh()
 }
 
 // Status sets the HTTP status for the response.
-// This method is chainable.
-func (ctx *Ctx) Status(status int) *Ctx {
-	ctx.Fasthttp.Response.SetStatusCode(status)
-	return ctx
+func (ctx *Context) Status(status int) {
```
Comment (ReneWerner87) [inline ctx.go:1219]: why did you remove this return value here ? is this not implementable in the interface ?

File: ctx.go (hunk 1 above)
Comment (thomasvvugt) [reply ctx.go]: I made it private, however AcquireCtx() now return the unexported type of this struct.

File: ctx.go (hunk 3 above)
Comment (thomasvvugt) [reply ctx.go]: I re-added the chainability of these methods by using the interface

**Comment by thomasvvugt** [comment]
Interface types can store the copy of a value, or a value can be shared with the interface by storing a pointer to the value’s address. An example of using a working struct pointer to extend upon the `Ctx` struct in Fiber's current version would be:

```go
type Custom struct {
    *fiber.Ctx
}

func (c *Custom) SendString(body string) {
    c.Send("Overriden, sorry 👋\n\n")
}

func main() {
    app := fiber.New()

    app.Get("/", func(c *fiber.Ctx) {
        cc := &Custom{c}
        cc.SendString("Hello, World!")
    })

    log.Fatal(app.Listen(":3000"))
}
```

Some arguments to use/not use the interface approach in comparison to the pointer struct approach;
- The interface implementation decides the type of parameter (a pointer or a value) to pass to a method/function that accepts the interface as a parameter.
- Since both type of receivers are allowed, there could be varying implementations of the same interface, which could lead to possible runtime errors during type conversion.

Let me know what you guys think about both approaches.
